correctly treat pan/zoom events of overlapping axes

raphaelquast opened the pull request:
## PR Summary
This is intended to fix the treatment of pan/zoom events for overlapping axes (e.g. #22324 )

<details>
<summary> 🐍  code to generate the plot of the gif </summary>

```python
import matplotlib.pyplot as plt

def styleax(ax, bg, c, zorder, txtxy=(.5,.5)):
    # bg=False sets the patch to invisible
    # any other value keeps the patch AS IS
    
    if bg is False:
        ax.patch.set_visible(False)
        for _, s in ax.spines.items():
            s.set_edgecolor(c)
        ax.tick_params(axis='x', colors=c)
        ax.tick_params(axis='y', colors=c)

    else:
        ax.patch.set_facecolor(c)
        for _, s in ax.spines.items():
            s.set_edgecolor(c)

        ax.tick_params(axis='x', colors=c)
        ax.tick_params(axis='y', colors=c)  
        
    ax.set_zorder(zorder)
    if txtxy:
        ax.text(*txtxy, 
                f"{'NO' if bg is False else ''} patch axes \n at zorder={zorder}", 
                c=c, horizontalalignment="center", verticalalignment="center", 
                bbox=dict(facecolor='w', alpha=0.75),
                fontsize=7)

f, ax = plt.subplots()
styleax(ax, None, ".8", 0, False)
styleax(ax.twinx(), None, ".8", 0, False)  # a basic twinx

# manually added axes with background
ax = f.add_axes((.5,.5,.4,.4))
styleax(ax, None, "g", 0, (.7, .5))

# manually added axes with background and twinx
ax = f.add_axes((.5,.5,.2,.2))
styleax(ax, None, "b", 0)
styleax(ax.twinx(), None, "b", 0, False)

# manually added axes with NO background and twinx
ax = f.add_axes((.2,.25,.2,.2))
styleax(ax, False, "r", 0)
styleax(ax.twinx(), None, "r", 0, False)

# manually added axes with NO background on different zorder
ax = f.add_axes((0.05,.1,.55,.7))
styleax(ax, False, "c", 2, (.5,.7))
styleax(ax.twinx(), False, "c", 2, (.5,.7))

# manually added axes with background and twinx on different zorder
ax = f.add_axes((.5,.25,.2,.2))
styleax(ax, None, "m", 5)
styleax(ax.twinx(), None, "m", 5, False)
# a joined axes
ax2 = f.add_axes((.8,.25,.15,.15))
styleax(ax2, None, "m", 5)
ax2.sharex(ax)
```

</details>


![mpl_issue_2](https://user-images.githubusercontent.com/22773387/151589404-7e540830-3f82-414a-807f-29391c81c2c5.gif)


#### changes on pan/zoom events
- instead of triggering all axes, only the topmost ax with `ax.patch.get_visible = True` triggers pan/zoom events
- axes with `ax.patch.get_visible = False` pass events to axes below (e.g. at lower or equal zorder)
#### other changes
- to allow a coherent treatment for twin-axes, the `zorder` property is set to the `zorder` of the parent axes.
  in `ax._make_twin_axes()`


## PR Checklist

<!-- Please mark any checkboxes that do not apply to this PR as [N/A]. -->
**Tests and Styling**
- [x] Has pytest style unit tests (and `pytest` passes).
- [x] Is [Flake 8](https://flake8.pycqa.org/en/latest/) compliant (install `flake8-docstrings` and run `flake8 --docstring-convention=all`).

**Documentation**
- [x] New features are documented, with examples if plot related.
- [ ] New features have an entry in `doc/users/next_whats_new/` (follow instructions in README.rst there).
- [x] API changes documented in `doc/api/next_api_changes/` (follow instructions in README.rst there).
- [x] Documentation is sphinx and numpydoc compliant (the docs should [build](https://matplotlib.org/devel/documenting_mpl.html#building-the-docs) without error).

<!--
Thank you so much for your PR!  To help us review your contribution, please
consider the following points:

- A development guide is available at https://matplotlib.org/devdocs/devel/index.html.

- Help with git and github is available at
  https://matplotlib.org/devel/gitwash/development_workflow.html.

- Do not create the PR out of main, but out of a separate branch.

- The PR title should summarize the changes, for example "Raise ValueError on
  non-numeric input to set_xlim".  Avoid non-descriptive titles such as
  "Addresses issue #8576".

- The summary should provide at least 1-2 sentences describing the pull request
  in detail (Why is this change required?  What problem does it solve?) and
  link to any relevant issues.

- If you are contributing fixes to docstrings, please pay attention to
  http://matplotlib.org/devel/documenting_mpl.html#formatting.  In particular,
  note the difference between using single backquotes, double backquotes, and
  asterisks in the markup.

We understand that PRs can sometimes be overwhelming, especially as the
reviews start coming in.  Please let us know if the reviews are unclear or
the recommended next step seems overly demanding, if you would like help in
addressing a reviewer's comments, or if you have been waiting too long to hear
back on your PR.
-->

QuLogic commented on lib/matplotlib/backend_bases.py:
```diff
@@ -3058,14 +3058,50 @@ def press_pan(self, event):
+
+        # get all relevant axes and group them by patch-visibility and zorder
+        all_axes = defaultdict(lambda: defaultdict(list))
+        [all_axes[a.patch.get_visible()][a.zorder].append(a)
+         for a in self.canvas.figure.get_axes()
+         if a.in_axes(event) and a.get_navigate() and a.can_pan()]
```
Definitely do not use list comprehensions for their side effect.

raphaelquast replied:
ok, fair point, using a normal loop now!

QuLogic commented on lib/matplotlib/backend_bases.py:
```diff
@@ -3058,14 +3058,50 @@ def press_pan(self, event):
-                if a.in_axes(event) and a.get_navigate() and a.can_pan()]
-        if not axes:
+
+        # get all relevant axes and group them by patch-visibility and zorder
+        all_axes = defaultdict(lambda: defaultdict(list))
+        for a in self.canvas.figure.get_axes():
+            if a.in_axes(event) and a.get_navigate() and a.can_pan():
+                all_axes[a.patch.get_visible()][a.zorder].append(a)
```
So I'm wondering why we need to group the Axes? Can we not sort by reverse zorder, and pan all Axes down until the one that has a visible patch?

raphaelquast replied:
yes, in the first place that would be the same... however in a later stage I need the groups for 2 things:
- to trigger all other transparent axes with higher or equal zorder
- to check if an axes is a twin of a visible-patch axes at the same zorder (it was the only way I could find to avoid triggering twin-axes that are at the same zorder but whose parent is effectively below the clicked axes)

... Also, I don't think that making groups here should be a performance issue

jklymak commented:
This is a relatively large behaviour change.  Perhaps for the better, and I'd definitely support this as a default. However, is there a way for users to get the old behaviour and still keep the axes transparencies as they were?  It seems to me that we need a manual toggle for all of this?  

Perhaps a contrived example, but it's possible I make a plot and know that a useful portion of the main axes is covered by the inset and hence zoom over the inset expecting the main axes to re-center.  It is also quite likely that an inset is showing a fixed portion of the main axes and I don't _want_ to allow a zoom on it, in which case the zoom may as well pass through.  

I'm not particulalry opposed to this solution, but want to make sure it is though through.

raphaelquast commented:
@jklymak  of course... I tried to give this as much thought as i could 😄 

So the old behaviour would mean that irrespective of background and zorder, you always trigger all axes... not sure anyone really wants to go back to that.

- If I really want the inset-axes to have a background AND trigger other axes (whether they are below or not), using `sharex=ax` and/or `sharey=ax` would be the natural way to achieve this. 

- NOT triggering an axes and just passing the zoom events to axes below is not possible in the current behaviour and it is still not possible.   (I guess implementing it would indeed require an additional flag for axes... such as `disable_zoom` or `disable_pan` or something like that)

jklymak commented:
I guess I'm in favour of a reasonable default as proposed here, but also an explicit state on the Axes that can be queried and set.  Probably pan needs to be included?   

Certainly I don't think a zoom in one axes need be tied to anything on the other axes.  We just encouraged that functionality in an external package for now.

anntzer commented on lib/matplotlib/backend_bases.py:
```diff
@@ -3058,14 +3058,50 @@ def press_pan(self, event):
+        all_axes = defaultdict(lambda: defaultdict(list))
+        for a in self.canvas.figure.get_axes():
+            if a.in_axes(event) and a.get_navigate() and a.can_pan():
+                all_axes[a.patch.get_visible()][a.zorder].append(a)
```
Not commenting on the overall strategy; just on the current implementation...

The use of defaultdict and of `all_axes[True]` / `all_axes[False]` below is not so easy to follow.  I would just do something like (untested, and perhaps further simplifications are possible) `visible_axes_by_z = {}; invisible_axes_by_z = {}` then
```python
for ax in get_axes():
    if ax.get_visible(): visible_axes_by_z.setdefault(ax.zorder, []).append(ax)
    else: invisible_axes_by_z.setdefault(ax.zorder, []).append(ax)
```
(one problem with defaultdict is that *any* `__getitem__` could be inserting a new key into the dict, and it's hard to follow which ones do so; using setdefault makes this explicit.)

raphaelquast replied:
OK, I can change that... I thought that using a defaultdict should be fine in here since it stays inside the function and using it makes the code more compact. (but I agree that if you're not used to them they can be somewhat confusing...)

raphaelquast replied:
@anntzer I've now updated the implementation to use explicit `dict.setdefault(key, []).append()` and `dict.get(key, [])` instead of a `defaultdict`

raphaelquast commented:
@jklymak 
... I've once tried to implement a shared zoom functionality for `cartopy` myself, but I stopped since it can get quite messy with geo-axes at different projections...

you mean a state-variable for disabling zoom&pan on an axes right? I could add it, but I'd need someone to point me to the right location for that.  (also I don't know what your naming-conventions for such a variable would be)   Maybe it's better to open an issue and do this in another pull-request?

jklymak commented:
My personal opinion is we should figure out the api for allowing the user to control the pass through before we change the behaviour in a way the user can't get back to the previous behaviour.  We would only do that if we were sure the old behaviour was definitely wrong.

jklymak commented:
So I think the API would be:

`axes._captures_navigation`

- `None`: default as proposed here, and sentinel that the user hasn't explicitly set this. 
- `True`: stop further propagation down axes list.
- `False`: allow event to continue down the stack (current behaviour). 

Then we just need a `set/get_captures_navigation` pair of methods.  

This would live on `axes/_base.py`

raphaelquast commented:
Hey, 
I'd be willing to implement this, but I have 2 general points here that I think might be worth considering beforehand:
(I'm not a matplotlib dev but I hope my opinion is welcome...)

- First, if a new attribute is introduced, I'd suggest that it also directly allows to completely disable navigation on an axis (and then automatically forward the events to axes below)

  - maybe a more general `ax._navigation` attribute that accepts values like: [`None`, `"forward"`, `"capture"`, `"disable"`] ?

- Second, I'm not sure if the "continue down the stack" option is really useful at all (not even for transparent axes)
  (for twinned-axes yes, but they can always be handled explicitly)
  Note... events are not "continued" with respect to data-coordinates, the "continuation" only propagates the mouse-position to the other axes!
  
  ... in other words... here's what the "continue down the stack" option does with axes that are **not the same size**:
  
  - when you try to zoom-in on a rectangle in an inset-axes, you simultaneously zoom-in to the very same rectangle on all other axes (which is an area effectively located **behind the inset-axes** !) 
  - the effective size of the zoomed-in area in the "continued-axis" however is dependent on the relative size between the "continued-axis" and the axis the user selected as basis for his mouse-movement  (which intuitively is always the topmost axes)
    - the very same scaling issue happens also if you execute a "pan" on the inset-axes

... here's what I mean as a gif:

<details>
<summary> 🐍 code to reproduce </summary>

```python
import matplotlib.pyplot as plt
import numpy as np
data = np.mgrid[0:30, 0:30]

f, ax = plt.subplots()
ax2 = f.add_axes((.25,.25,.5,.5))
ax.scatter(*data, c="C0")
ax2.scatter(*data, c="C1")
```

</details>

![mpl_issue_3](https://user-images.githubusercontent.com/22773387/151774572-6697bb39-892e-47d4-a315-635b90cecfb1.gif)

jklymak commented:
> Note... events are not "continued" with respect to data-coordinates, the "continuation" only propagates the mouse-position to the other axes!

Thats understood, but replacing an explicit (though perhaps undesirable) behaviour with an implicit behaviour depending on murky heuristics is how we get into trouble unless we allow the user an escape hatch to specify things explicitly.  Given that no one has complained about the existing behaviour until now (that I know of) makes me a bit skeptical that we can just discount the old behaviour as a bug.

anntzer commented:
Just a quick drive-by comment: there's already `ax.set_navigate(False)`, which can disable pan/zoom on an Axes.

raphaelquast commented:
@anntzer thx, don't know how I've missed that...  (I've indicated the function in the `See Also` section of the `ax.set_captures_navigation` doc) 

@jklymak OK, I get your concerns here... I've now implemented the API as suggested!

Note however that there's not yet a way to get back to the current behavior globally... so at the moment, if you really need the old behavior, you'd have to set the navigation-capture mode for each axes individually...

Should `axes.captures_navigation=None` be ingested into `rcParams` so that users can override the global default?

jklymak commented:
> Should `axes.captures_navigation=None` be ingested into `rcParams` so that users can override the global default?

I wouldn't unless there is an outcry.  To be clear, I support the default behaviour, and think it is what most people would prefer, so I think its OK to ask the _few_ folks who may object to set the state manually on each axes.

raphaelquast commented:
@jklymak OK, sounds good to me! ...then I think from my side I did what I can to push this forward...
(I have no clue why the docs are failing now, but it looks more like a CI related issue than something connected to the changes here)

jklymak commented:
Yes, if you rebase, the docs should build again.  Setuptools broke us.

jklymak commented:
BTW: this definitely needs an API behaviour change note.  `/docs/api/next_api/behavior`

timhoffm commented on lib/matplotlib/axes/_base.py:
```diff
@@ -585,6 +586,9 @@ def __init__(self, fig, rect,
             Set a fixed aspect for the Axes box, i.e. the ratio of height to
             width. See `~.axes.Axes.set_box_aspect` for details.
 
+        captures_navigation : bool, optional
```
Can we make this more explicit, e.g. `bool or 'transparent_axes'` Or even three strings?

Also, use imperative style `capture_navigation`; c.f. `sharex`.

More general naming: Is capture an established technical term in this context? I'd say one captures events not navigation. If events is too broad, we can afford the longer "navigation events" for this specialized functionality. But OTOH, should this be limited to zoom/pan? IMHO e.g. also picking should follow the same logic.

raphaelquast replied:
i mark this as resolved since the discussion on default args continued below

raphaelquast commented on lib/matplotlib/axes/_base.py:
```diff
@@ -585,6 +586,9 @@ def __init__(self, fig, rect,
             width. See `~.axes.Axes.set_box_aspect` for details.
 
+        captures_navigation : bool, optional
+            Control how pan/zoom events are forwarded to other axes.
```
Suggestion implemented

raphaelquast commented:
> BTW: this definitely needs an API behaviour change note. `/docs/api/next_api/behavior`

@jklymak   I added a note to `/next_api_changes/behavior/22347-RQ.rst`

raphaelquast commented:
hey, before this gets lost in time... 
Anything else you'd need from my side to merge this?  (I tried another rebase but somehow the docs still don't build)

oscargus commented:
Try getting rid of leading spaces on lines 9 and 10. That should solve the doc problem.

(For the PR itself, I cannot really say.)

QuLogic commented:
I believe the doc failure is that you need to add your new functions to `doc/api/axes_api.rst`.

raphaelquast commented:
hey all, 
so I did another rebase and implemented all suggestions but to be honest, I really have no clue why the docs still fail...
would be nice if anyone from your side could have a look at this because I think aside of the docs it's ready

timhoffm commented on doc/api/next_api_changes/behavior/22347-RQ.rst:
```diff
@@ -0,0 +1,13 @@
+The forwarding of pan/zoom events is now determined by the visibility of the
+background-patch (e.g. ``ax.patch.get_visible()``) and by the ``zorder`` of the axes.
+
+- Axes with a visible patch do not forward events to axes below.
+  - Only the axis with the highest ``zorder`` that contains the event is triggered
+    (if there are multiple axes with the same ``zorder``, the last added axis counts)
+- Axes with an invisible patch forward events to axes at lower or equal ``zorder``
```
> /home/circleci/project/doc/api/next_api_changes/behavior/22347-RQ.rst:9: ERROR: Unexpected indentation.

reStructuredText does not like this formatting. I think l.8-9 do not have to be a bullet point:

```suggestion
- Axes with a visible patch do not forward events to axes below.
  Only the axis with the highest ``zorder`` that contains the event is triggered
  (if there are multiple axes with the same ``zorder``, the last added axis counts)
- Axes with an invisible patch forward events to axes at lower or equal ``zorder``
```

If you really want in indented sublist, you'd have to add blank lines around l.8-9 otherwise ReST doesn't understand it.

raphaelquast replied:
thanks! lets give it a try 🙂

timhoffm commented on doc/api/next_api_changes/behavior/22347-RQ.rst:
```diff
@@ -0,0 +1,13 @@
+Correctly treat pan/zoom events of overlapping axes
```
```suggestion
Correctly treat pan/zoom events of overlapping Axes
```

We've decided to captialize this, when referring to the Axes (class) as a concept so that this is not confused with the plural of "axis". See #18726. Should also be changed for all other "Axes" mentionings below.

timhoffm commented on doc/api/next_api_changes/behavior/22347-RQ.rst:
```diff
@@ -0,0 +1,13 @@
+    (if there are multiple axes with the same ``zorder``, the last added axis counts)
+- Axes with an invisible patch forward events to axes at lower or equal ``zorder``
+
+To override the default behavior and explicitly set whether an axis
```
```suggestion
To override the default behavior and explicitly set whether an Axes
```

timhoffm commented on doc/api/next_api_changes/behavior/22347-RQ.rst:
```diff
@@ -0,0 +1,13 @@
+
+- Axes with a visible patch do not forward events to axes below.
+  - Only the axis with the highest ``zorder`` that contains the event is triggered
+    (if there are multiple axes with the same ``zorder``, the last added axis counts)
```
```suggestion
    (if there are multiple Axes with the same ``zorder``, the last added Axes counts)
```

timhoffm commented on lib/matplotlib/axes/_base.py:
```diff
@@ -597,6 +598,10 @@ def __init__(self, fig, rect,
             Set a fixed aspect for the Axes box, i.e. the ratio of height to
             width. See `~.axes.Axes.set_box_aspect` for details.
 
+        capture_navigation : bool or None,
```
```suggestion
        capture_navigation_events : bool or None,
```
I feel this is a bit clearer. - Should then also used for the private variable and the setter/getter.

raphaelquast replied:
✔️ implemented!

timhoffm commented on lib/matplotlib/axes/_base.py:
```diff
@@ -4601,3 +4614,28 @@ def get_shared_x_axes(self):
+
+            - True: Events are only executed on this axes.
+            - False: Forward events to other axes with lower or equal zorder
+            - None: Default behaviour ("True" for axes with an invisible patch
```
`set_capture_navigation_events(None)` is quite opaque. Could we have a more telling name than None? There's precedence for a bool+str tristate e.g. in https://matplotlib.org/devdocs/api/_as_gen/matplotlib.axes.Axes.set_axisbelow.html?highlight=set_axisbelow#matplotlib.axes.Axes.set_axisbelow.

Something like this is understandable:
`set_capture_navigation_events('if-axes-is-non-transparent')`

Maybe you can find a shorter or better text?

raphaelquast replied:
Hey @timhoffm, I can implement this but I'd very much like an agreement first to avoid having to change it multiple times...
Here are some suggestions from my side:
- `"axes_bg_visible"`, `"if_axes_bg_visible"`
- `"visible_background"`, `"visible_bg"`, `"background_visible"` or `"bg_visible"` 
  (it's a method of an axes... so it should be clear this refers to the axes)

I'm happy with either one (even your suggestion... however I find it a bit long and it's not about the visibility of the axes but about the visibility of the background patch of the axes).

raphaelquast replied:
hey @timhoffm ... just checking in to make sure it keeps rolling 😅

timhoffm replied:
Sorry, this got lost in the stream of notifications. Thanks for nudging.

I agree that "axes" can be implied. I suggest `capture_navigation_events='if_background_visible'`.

Yes, it's a bit long, but it's quite self-explanatory now. And it's rare that users have to write it themselves because it's the default.

tacaswell replied:
Is `'auto'` sufficient?

raphaelquast replied:
`"auto"` is now implemented ...

timhoffm reviewed (commented):
Thanks for keepting this rolling :+1:

QuLogic commented on doc/api/next_api_changes/behavior/22347-RQ.rst:
```diff
@@ -0,0 +1,13 @@
+background-patch (e.g. ``ax.patch.get_visible()``) and by the ``zorder`` of the axes.
+
+- Axes with a visible patch do not forward events to axes below.
+  Only the axis with the highest ``zorder`` that contains the event is triggered
+  (if there are multiple axes with the same ``zorder``, the last added axis counts)
```
They are all `Axes`:
```suggestion
  Only the Axes with the highest ``zorder`` that contains the event is triggered
  (if there are multiple Axes with the same ``zorder``, the last added Axes counts)
```

QuLogic commented on doc/api/next_api_changes/behavior/22347-RQ.rst:
```diff
@@ -0,0 +1,13 @@
+- Axes with an invisible patch forward events to axes at lower or equal ``zorder``
+
+To override the default behavior and explicitly set whether an Axes
+should capture navigation events, use ``ax.set_capture_navigation_events()``.
```
Should link:
```suggestion
should capture navigation events, use `.Axes.set_capture_navigation_events`.
```

QuLogic commented on lib/matplotlib/axes/_base.py:
```diff
@@ -597,6 +598,10 @@ def __init__(self, fig, rect,
             Set a fixed aspect for the Axes box, i.e. the ratio of height to
             width. See `~.axes.Axes.set_box_aspect` for details.
 
+        capture_navigation_events : bool or None,
```
```suggestion
        capture_navigation_events : bool or None
```

QuLogic commented on lib/matplotlib/axes/_base.py:
```diff
@@ -4601,3 +4614,28 @@ def get_shared_x_axes(self):
+        self._capture_navigation_events = capture
+
+    def get_capture_navigation_events(self):
+        """Get how pan/zoom events are forwarded to other axes."""
```
```suggestion
        """Get how pan/zoom events are forwarded to Axes below this one."""
```

QuLogic commented on lib/matplotlib/axes/_base.py:
```diff
@@ -4601,3 +4614,28 @@ def get_shared_x_axes(self):
+
+    def set_capture_navigation_events(self, capture):
+        """
+        Set how pan/zoom events are forwarded to other axes.
```
```suggestion
        Set how pan/zoom events are forwarded to Axes below this one.
```

QuLogic commented on lib/matplotlib/backend_bases.py:
```diff
@@ -3058,14 +3058,62 @@ def press_pan(self, event):
+        axes_capture = {True: {}, False: {}}
+        for a in self.canvas.figure.get_axes():
+            if a.in_axes(event) and a.get_navigate() and a.can_pan():
+                # By default, axes with a transparent patch forward events
+                # while axes with an intransparent patch capture events.
```
Why is this resolved? The change hasn't been made.

QuLogic commented on lib/matplotlib/backend_bases.py:
```diff
@@ -3060,14 +3060,62 @@ def press_pan(self, event):
-                if a.in_axes(event) and a.get_navigate() and a.can_pan()]
-        if not axes:
+
+        # Get all axes and group them by their navigation-capture behaviour.
+        axes_capture = {True: {}, False: {}}
```
I'm still not certain why this is necessary. Why can't you sort by zorder and stop at the first Axes that captures?

raphaelquast replied:
hey, so I tried to explain this before very briefly here: https://github.com/matplotlib/matplotlib/pull/22347#discussion_r795041340

I'll try to elaborate a bit more on this... the main reason why I implemented it is the following:
- assume you have an axis (say `ax1`) with a visible patch (e.g. a capturing axis)
- and a (maybe but not necessarily overlapping) axis (`ax2`) that also has a twin-axis (`ax2_twin`)

now, if you pan/zoom `ax1`, the following happens:
- if `ax2` is **at the same or higher zorder**, `ax2_twin` would always be triggered because `ax2_twin` has **no visible patch** (e.g. it's non-capturing just as any other twin-axis).
- to avoid this, the implemented logic is the following:
  - first trigger the topmost (capturing) axes (e.g. in `axes_capture[True][zorder]`) and all twin-axes of it  
    (those ones we 100% always want to trigger)
  - then, check for non-capturing axes (e.g. no visible patch) at the same or higher zorder  
    (`axes_capture[False][>=zorder]`)
    - if an axis is actually just a twin-axis of an axis with a visible patch (e.g. its parent is in `axes_capture[True]`)  **don't trigger it** (note that we would have triggered it already as sibling of it's parent if relevant! )
    - otherwise trigger it (if it has not already been triggered)

So far this was the most easy way I could think of to ensure triggering only axes we really want to trigger...

tacaswell replied:
I have some concerns about the "same or higher zorder" part of this.  In all other cases we sort on zoreder using a stable sort so we always fully break the degeneracy and based on the order the Artists were added.  That is to say, even though two Axes (or any artist) may have identical zorder's there is a cannonical ordering and one is ahead of the other. 

I think the case that is prolemaatic here is `ax2_twin > ax1 > ax2` (that is ax1 is "between" the two layers of of the twin axes and you want to trigger all three) vs `ax2_twin > ax2 > ax1` and `ax1 > ax2_twin > ax2` where you only want to trigger ax1 or the ax2 set?

I am also very late to the party here and have not fully read up so I could me missunderstanding something....

raphaelquast replied:
@tacaswell 😄 I tried to capture as many possibilities as possible in the test-plot... 
concerning your question, twin-axes are treated explicitly so all mentioned cases are already properly handled!

tacaswell commented:
@raphaelquast I took the liberty of rebasing this.

tacaswell commented:
@raphaelquast Would you be willing to squash this down to a handful of commits?

tacaswell commented:
I am 👍🏻 on merging this if:

 1. squash to fewer (1?) commit
 2. use something other than `None` as the "do what I mean" logic of forwarding based on patch visibility.

raphaelquast commented:
@tacaswell since this was my first attempt to squash commits I of course made some errors and now there are even more commits... sorry 😅 

To avoid spending too much time fixing my self-made git rebase/squash issues I created a new cleaned-up and squashed pull-request #24631   (hope that's OK for you ... if yes, this PR can be closed and all discussions can be moved to the new one)

- I've implemented using `ax.capture_navigation_events = "auto"` as the default (instead of `None`)
  (and updated docs etc. accordingly)
- squashed everything to a single commit

-----

... also, to also answer your remaining question:
> I think the case that is problematic here is ax2_twin > ax1 > ax2 (that is ax1 is "between" the two layers of of the twin axes and you want to trigger all three) vs ax2_twin > ax2 > ax1 and ax1 > ax2_twin > ax2 where you only want to trigger ax1 or the ax2 set?

This is already addressed by the fact that twin-axes are handled explicitly irrespective of any other logic!
  - ax2_twin > ax1 > ax2:
    -  since ax2_twin is a twin-axes of ax2, it will always trigger ax2 as well independent of what's in between them
  - ax2_twin > ax2 > ax1:
    - if ax2 has a visible patch, ax1 will not trigger
  - ax1 > ax2_twin > ax2
    - if ax1 has a visible patch, ax2 and its twin will not trigger

tacaswell commented:
Just to keep the discussion tidy (and to make it easier to find this discussion later when blame lands on a given commit), I have a slight preference to merge it via this PR.

```
git push --force-with-lease YOUR_REMOTE_NAME nested_ax_zoom__candidate:nested_ax_zoom

```

is the git incantation that you need.  

Git is trying to protect you from accidentally orphaning commits on the server so when you did the push it probably said "I can't do that, try `git pull` and then try again"?  When you do that you effectively bring all of the old history back into your new history!  To get around this `git push` has a `--force` flag which does no checks, however it is basically blind.  `--force-with-lease` does the force push only if the remote branch is where you think it is (which saves me about once a month!).

raphaelquast commented:
@tacaswell yes, that's perfectly understandable!

Thanks for the explanation! much appreciated!   (now all is properly squashed to 1 commit)

tacaswell commented on lib/matplotlib/axes/_base.py:
```diff
@@ -4065,7 +4072,12 @@ def set_navigate(self, b):
         Parameters
         ----------
-        b : bool
+        b : bool or None
```
```suggestion
        b : bool or 'auto'
```

tacaswell replied:
nvm, I think this is wrong.

tacaswell commented:
One final thing, can you put the code at the top in the showcase gallery?  I do not think we need to go so far as to include the gif, but having that code readily available is probably good as in this case seeing is believing.

raphaelquast commented:
yes I can do that but you need to point me to the right location...   (I'll do a final squash when everything is done)

raphaelquast commented:
@tacaswell OK, I brushed up the example a bit and added it to `examples/showcase`.
(and I did another squeeze so from my side I guess all is done)

ksunden commented:
Doc building warnings do appear to be related

raphaelquast commented:
@tacaswell @ksunden  I had another look and found the typo that caused the docs to fail (used `capture_navigation_event` instead of `capture_navigation_events` in `axes_api.rst`) ... now all checks pass!

I did a rebase with `main` and squashed all into 1 commit again. 
From my side I think this is now finally ready to be merged 🥳

jklymak reviewed (commented):
These are maybe a pain to test, but quite a few of the cases don't have test coverage.

raphaelquast commented on lib/matplotlib/backend_bases.py:
```diff
@@ -3182,14 +3182,62 @@ def press_pan(self, event):
+                        axes_capture[False].setdefault(
+                            a.get_zorder(), []).append(a)
+                else:
+                    axes_capture[a.get_capture_navigation_events()].setdefault(
```
I only need to check for `"auto"` because in any other case the capturing behavior is given by the value of `a.get_capture_navigation_events()` (e.g. to provide the requested ability to manually override the capturing behavior)

... or do you mean an explicit test that the value must be one of [True, False, "auto"] ? (If yes, I'd rather put the check in `a.set_capture_navigation_events()` directly)

QuLogic replied:
I believe @jklymak is talking about unit tests, not the test condition for the `if`, because codecov says this line is untested.

tacaswell commented on lib/matplotlib/axes/_base.py:
```diff
@@ -616,6 +617,11 @@ def __init__(self, fig,
 
+        capture_navigation_events : bool or "auto", default: "auto"
+            Control whether pan/zoom events are passed through to Axes below
+            this one. "auto" is *True* for axes with an invisible patch and *False* 
+            otherwise.
```
```suggestion
            this one. "auto" is *True* for axes with an invisible patch and
            *False* otherwise.
```

tacaswell commented:
https://github.com/matplotlib/matplotlib/blob/9f3c6f4e29d810ffb8623308222244c7ecdd72ad/lib/matplotlib/tests/test_backend_bases.py#L201-L259 is probably a template for how to test this.

tacaswell commented:
I am still inclined that we should merge this for 3.7 based on interactive testing and maybe minimal unit testing and open an issue to add exhaustive testing.

raphaelquast commented:
Damn... I was too intrigued and opened pandoras box 😅  
I started writing a basic test (based on the suggested example of @tacaswell)  and in the process I unfortunately realized that there is in fact still important work do...

While I would have loved to get this done soon, for the sake of completeness I now think that it would be best to at least address the remaining points I found before merging.

To summarize:

1. zoom events need the same treatment as pan events!
    - to clarify... I mean using the "zoom-to-rectangle" tool
    - at the moment this PR only addresses using the pan-tool (to pan or zoom)
    - this shouldn't be too much effort since it's basically just copying the procedure of `.press_pan()` to `.press_zoom()`
2. It seems that there might be a race-condition somewhere when using `ax.sharex()`...
    - in this case, the test starts to fail or succeed randomly and some axes are mysteriously triggered twice
      (when using only ordinary axes and twin-axes (e.g. created with `ax.twinx()` all is fine))
    - I haven't had enough time to really figure out what's going on here... 

Let me know what you think!

tacaswell commented:
> Let me know what you think!

Given that you nerd sniped your self [1], lets push this to 3.8 rather than try to rush it in the next week and a half.


[1] https://xkcd.com/356/

tacaswell commented:
To clarify if you do get through that list by the time we branch 3.7.x 👍🏻 on it going in!  I just moved it out of the 3.7 milestone to make it easier to manage the final burn down of open PRs.

raphaelquast commented:
Well.. I know it's a pity, but I think it's for the better... (capturing pan-events and non-capturing zoom-events would be quite a source for confusion)

I'll see what I can do 🚀

raphaelquast commented:
@tacaswell 

I've implemented the capture-logic for zoom events and added a unittest to check the basic functionality.
Now everything checks out fine.

The only thing I'm still unsure about is the problem with shared axes. 
If you define `ax_t_twin` and `ax_b_twin` with `ax.sharex()` the tests start to randomly fail
and this warning shows up:

```python
D:\python_modules\matplotlib\lib\matplotlib\tests\test_backend_bases.py:483: UserWarning: Attempting to set identical low and high ylims makes transformation singular; automatically expanding.
  tb.release_zoom(stop_event)
D:\python_modules\matplotlib\lib\matplotlib\axes\_base.py:4196: RuntimeWarning: divide by zero encountered in double_scalars
  factor = (symax0 - symin0) / (symax - symin)
D:\python_modules\matplotlib\lib\matplotlib\axes\_base.py:4198: RuntimeWarning: invalid value encountered in double_scalars
  symax1 = symax0 + factor * (symax0 - symax)
```


Any ideas what's happening here?
I expected `ax.sharex(...)` to behave similar to `ax.twinx()` ?


https://github.com/matplotlib/matplotlib/blob/9ad477d8e9433593f09a65e4926c18f73179ce4a/lib/matplotlib/tests/test_backend_bases.py#L414-L430

jklymak commented:
.. moving to draft doesn't mean the PR can't proceed, just that its not currently reviewable.  However, if thats not true (you need a review) feel free to move back out of draft...

raphaelquast commented:
@jklymak I think I found the issue I had... and I'd very much like to get a review on this!

The issue was that I explicitly handled shared axes and twinned axes... however shared axes share axes-limits no matter what, 
so the additional trigger caused false limits. (Also because the click position would be wrong since it would require the transformed position relative to the axes in order to properly maintain limits... )

concerning the tests... circleci is complaining about installation errors now but that's not related in any way to this PR
```
  Beginning with Matplotlib 3.8, Python 3.9 or above is required.
  You are using Python 3.8.16.
```

ksunden commented:
A rebase should get the newer Circle config and then work (it gets the config from the PR branch, but tries to build on the merged into main version)

raphaelquast commented:
@ksunden I did a rebase, now only `appveyor` complains... but again for a test named ` def test_sphinx_gallery_example_header():` which should not be related to this PR.

ksunden commented:
Is it possible for you to rebase instead of merging? https://matplotlib.org/devdocs/devel/development_workflow.html#rebasing-on-upstream-main

raphaelquast commented:
oh how i hate rebasing 😅 
anyways... now its done and I also squashed everything down to 1 commit again!

raphaelquast commented on lib/matplotlib/axes/_base.py:
```diff
@@ -4599,3 +4613,28 @@ def draw(self, renderer):
 
     return _MinimalArtist(figure, artists).draw(renderer)
+
+    def set_capture_navigation_events(self, capture):
```
woha... that must have happened during my fight with the rebase... sorry i could have found that myself... fixing it right away!

raphaelquast commented:
Hey all, before this gets lost in time, I'll give it another try...

- What's still missing to get this merged?
  - are there general concerns on the implementation?
  - functionalities not working as expected?

From my side, the tests are now quite general and all works as expected.

greglucas commented on lib/matplotlib/backend_bases.py:
```diff
@@ -3218,14 +3218,71 @@ def press_pan(self, event):
+        axes_capture = {True: {}, False: {}}
+        for a in self.canvas.figure.get_axes():
+            if a.in_axes(event) and a.get_navigate() and a.can_pan():
+                # By default, axes with an invisible patch forward events
```
```suggestion
                # By default, events pass through invisible patches
```
I'm a bit confused by the wording choice of "forward". That word leaves it a bit ambiguous to me whether this axes handles the event and also passes it on, or if it just passes it on. I _think_ you're implying the first case where events simply pass through? I think this could also be changed in docstrings if you agree.

greglucas commented on lib/matplotlib/backend_bases.py:
```diff
@@ -3218,14 +3218,71 @@ def press_pan(self, event):
+            if a.in_axes(event) and a.get_navigate() and a.can_pan():
+                # By default, axes with an invisible patch forward events
+                # while axes with a visible patch capture events.
+                if a.get_capture_navigation_events() == "auto":
```
```suggestion
                if a.get_capture_navigation_events() is None:
```

greglucas commented on lib/matplotlib/backend_bases.py:
```diff
@@ -3273,19 +3330,86 @@ def press_zoom(self, event):
+        if axes_capture[True]:
+            zorder = max(axes_capture[True])
+            ax = axes_capture[True][zorder][-1]
+            ax.start_pan(event.x, event.y, event.button)
```
```suggestion
            ax.start_zoom(event.x, event.y, event.button)
```

There seems to be a lot of code duplication between the pan and zoom, would it make sense to break some of it out into a private helper?

greglucas commented on doc/api/next_api_changes/behavior/22347-RQ.rst:
```diff
@@ -0,0 +1,13 @@
+The forwarding of pan/zoom events is now determined by the visibility of the
+background-patch (e.g. ``ax.patch.get_visible()``) and by the ``zorder`` of the axes.
+
+- Axes with a visible patch do not forward events to axes below.
+  Only the Axes with the highest ``zorder`` that contains the event is triggered
+  (if there are multiple Axes with the same ``zorder``, the last added Axes counts)
+- Axes with an invisible patch forward events to axes at lower or equal ``zorder``
```
```suggestion
- Axes with a visible patch capture the event and do not pass it on to axes below.
  Only the Axes with the highest ``zorder`` that contains the event is triggered
  (if there are multiple Axes with the same ``zorder``, the last added Axes counts)
- Axes with an invisible patch is also invisible to events and they are passed on to the axes below.
```

greglucas reviewed (commented):
Some quick feedback to consider. I didn't read the entire message chain, so I may have missed some context on why certain values/words were chosen already. But, my overarching comments that I put more detail inline about are:
* the word "forward" is confusing/ambiguous to me
* we usually use `None` as default instead of "auto"
* See if you can re-use your axes-calculating logic between the pan and zoom functions by splitting it into a private helper which would also keep those functions manageable size to see at a glance what is going on.
*

jklymak commented:
Sorry this is dragging @raphaelquast - your perseverance is very much appreciated.  I think you still have some actionable comments from Greg above.  He is usually great about re-reviews, so feel free to ping!

raphaelquast commented:
Hey @jklymak,
Thanks for your follow up on this one!   (and @greglucas, thanks for your comments!)
It's not easy to keep track on such long-term PRs but I think we're getting close...

I got a bit carried away by other tasks recently, but I'll do my best to find some time to finalize this as soon as possible.
(The good thing is, I do have a new motivation to fix this since the upcoming [eomaps v7](https://github.com/raphaelquast/EOmaps) release will allow zoomable inset-maps.)


### To summarize, the remaining steps are: 

**Someone** has to decide this once and for all:
- [ ] agree on default arguments for `get_capture_navigation_events` (`"auto"` or `None`?)
  - `None` seems to be the commonly used default, but in this case it also somewhat hides the fact that there is a special capture logic applied by default (e.g. based on the background patch visibility)
  - `auto` would be at least a little bit more descriptive (to avoid confusions with `False`)

In the end I think this is a very minor problem since almost nobody will tinker with this option anyways... 
My preference would be to stick to `"auto"` but I understand the need for consistency, so if you prefer `None`, I have no objections.

TODOs:
- [ ] refactor code to avoid duplications (e.g. handling of pan/zoom events is quite similar)
- [ ] simplify example

tacaswell commented:
I would lean to `"auto"` here, but we should put this on the agenda for next week.

jklymak commented:
Sorry for only drive-by reviewing this, but per https://github.com/matplotlib/matplotlib/pull/22347#issuecomment-1025784134 is this distinct enough from `set_navigate`?

raphaelquast commented:
@jklymak At the moment, `ax.set_navigate` turns off navigation completely and `ax.set_capture_navigation_events` handles relaying of navigation events to axes below.

It would be possible (and maybe actually preferable?) to supercharge `ax.set_navigate` with more input options instead of introducing a new method... something like:

- True: use "auto" capture behavior
- "capture": capture navigation events
- "relay" or "forward"? : relay navigation events
-  False: turn navigation off

jklymak commented:
Would that give you all the flexibility you need, solve the simplest case, and be relatively easy to explain to users?  If so, maybe that is a good way to go forward?

tacaswell commented:
Two comments from today's call:

1. change the same of `set_capture_events` to `set_forward_events` and invert the logic.   The docs explaining this start with the word "forward" so it is a hint!
2. could you write an "unrolled" version of example (in addition to kitchen sink example)?  The kitchen sink example is good for testing but not good for understanding.

raphaelquast commented:
Hey, thanks for the follow-up on this!  
(And sorry for not being able to join the meeting...)

I can try to implement this in the next few days, but I'll be afk for all of August, 
so not sure I manage to find enough time to finalize this beforehand.

1. OK I can do that!
2. Yes, I can do that as well, but I need some clarifications first... 
   - Which one is the kitchen sink example now?  
     The current (stripped down) one or the initial (extensive but a bit confusing) one?
   - Not really sure what you mean by an "unrolled" example?

greglucas commented on lib/matplotlib/backend_bases.py:
```diff
@@ -2977,6 +2977,66 @@ def _zoom_pan_handler(self, event):
+            else:
+                continue   # continue to lower zorder if we did not break
+
+            break          # break if we hit a capturing axes
```
This is a dangling break?

Or now that I'm staring at this a bit more, does the `continue` in the `else:` clause above actually continue the outer for-loop? Maybe I'm just not picking up on it immediately, but the for-else with a continue not applying to that for-loop could use an extra comment here I think just to clarify what is explicitly going on.

raphaelquast replied:
I understand that this is a rarely encountered construct... but in my opinion it's the easiest way to conditionally break out of nested loops...   I'll try to add more detailed comments to avoid confusion!

To clarify:  You are right, the `else: continue` is there to continue the outer loop only if no `break` was encountered in the inner loop. 

This way, if the inner loop breaks, the outer loop breaks as well, and if the inner loop finishes without an explicit break, the outer loop continues to the next level.

From the [docs](https://docs.python.org/3/tutorial/controlflow.html#break-and-continue-statements-and-else-clauses-on-loops): 
> *When used with a loop, the else clause has more in common with the else clause of a [try](https://docs.python.org/3/reference/compound_stmts.html#try) statement than it does with that of [if](https://docs.python.org/3/reference/compound_stmts.html#if) statements: a [try](https://docs.python.org/3/reference/compound_stmts.html#try) statement’s else clause runs when no exception occurs, and a loop’s else clause runs when no break occurs.*

raphaelquast replied:
I added a few more comments... please re-open this discussion in case you think my explanations are insufficient!

greglucas commented on lib/matplotlib/backend_bases.py:
```diff
@@ -2977,6 +2977,66 @@ def _zoom_pan_handler(self, event):
+        # call "ax.start_pan(..)" on all relevant axes of an event
+
+        def _ax_filter(ax):
+            return ax.in_axes(event) and ax.get_navigate() and ax.can_zoom()
```
```suggestion
            return ax.in_axes(event) and ax.get_navigate() and ax.can_pan()
```
pan?

If you are also using it in zoom below, perhaps you want to make this function take a `zoom` or `pan` argument as well to use it for both?

greglucas commented on lib/matplotlib/backend_bases.py:
```diff
@@ -3057,21 +3113,21 @@ def press_zoom(self, event):
-        if self._nav_stack() is None:
-            self.push_current()  # set the home button to this view
+
+        axes = self._start_event_axes_pan(event)
```
Do we need an `if not axes: return` here so we don't add the id_zoom below unnecessarily?

raphaelquast replied:
yes that makes sense... implemented!

greglucas reviewed (commented):
A few more comments where I was confused at some of the logic happening in the code. I think we are getting close.

One major issue is that right-click zoom (i.e. zoom out) does not work on the lower two axes of your example. I get: `ValueError: Axis limits cannot be NaN or Inf` (I also only get this in the "y-direction" when holding "x" everything seems to work as expected, but holding "y" throws the `ValueError`

ksunden commented:
Did you mean to close this?

raphaelquast commented:
@ksunden NO definitely not.. I actually wanted to finally finish it by implementing the last minor changes and merging all to 1 commit. 

However, in the process I somehow managed to mixup histories (still not really sure how that happened) and now GitHub claims that the branch-histories are unrelated  and automatically closed the history.

I fixed the mixup, but GitHub now won't let me re-open the PR even though all is fine now (opening a new PR works just fine).

Really sorry about this... I'm unsure how to proceed now...  (new PR is here #27148)

raphaelquast commented:
reopen automatic close..

raphaelquast commented:
@all, sorry for the confusion with a temporary close of the PR... 

I've now merged everything into 1 commit and I think this is finally ready to be merged!

For future reference: 
There is a bug concerning twinned axes of shared axes (e.g. they are not properly triggered) but this is **not** addressed by this PR (e.g. bug persists in the exact same way as before). 

This PR only addresses the handling of overlapping pan/zoom events.

raphaelquast commented:
Since GitHub keeps auto-closing this PR without any obvious reason, here's a new PR #27148

raphaelquast commented:
@ksunden 
A final comment on this since I've investigated what went wrong while I've tried to finalize and cleanup this PR...

Apparently, I ran into an GitHub issue (https://github.com/isaacs/github/issues/361) which caused this PR to be closed and also prevents re-opening this PR.

The events that led to this are the following
- I accidentally force-pushed unrelated history resulting in an automatic close of the PR on GitHub
- Once I realized, I undid all changes and force-pushed back to a clean state
- Now the [GitHub issue](https://github.com/isaacs/github/issues/361) triggers which prevens force-pushed PRs to be re-opened once they have been closed.
- As a consequence, there's now #27148 which is a new PR for the exact same branch (cause somehow GitHub is fine with a new PR but not with re-opening the old one)

Really sorry about this... I actually just wanted to finalize this in a clean way and messed up quite a bit 😅 

Not sure how we want to proceed with the duplicated PRs now... aside of that, this PR would be ready for a final review!
